README.md
[![GitHub](https://img.shields.io/github/license/axi0m/ratatoskr?color=bright-green&style=flat-square)](https://github.com/axi0m/ratatoskr/blob/main/LICENSE.md)
![Maintenance](https://img.shields.io/maintenance/yes/2022?style=flat-square)
[![Code Style Black](https://img.shields.io/badge/code%20style-black-000000.svg?style=flat-square)](https://github.com/psf/black)
![GitHub Pipenv locked Python version](https://img.shields.io/github/pipenv/locked/python-version/axi0m/ratatoskr?style=flat-square)
[![Python Build with poetry](https://github.com/axi0m/ratatoskr/actions/workflows/build-poetry.yml/badge.svg?style=flat-square)](https://github.com/axi0m/ratatoskr/actions/workflows/build-poetry.yml)
[![Python Build with pipenv](https://github.com/axi0m/ratatoskr/actions/workflows/build-pipenv.yml/badge.svg?style=flat-square)](https://github.com/axi0m/ratatoskr/actions/workflows/build-pipenv.yml)
[![Python Build with pip](https://github.com/axi0m/ratatoskr/actions/workflows/build-pip.yml/badge.svg?style=flat-square)](https://github.com/axi0m/ratatoskr/actions/workflows/build-pip.yml)
[![flake8 lint](https://github.com/axi0m/ratatoskr/actions/workflows/flake8-lint.yml/badge.svg?style=flat-square)](https://github.com/axi0m/ratatoskr/actions/workflows/flake8-lint.yml)
[![CodeQL](https://github.com/axi0m/ratatoskr/actions/workflows/codeql.yml/badge.svg)](https://github.com/axi0m/ratatoskr/actions/workflows/codeql.yml)

# Ratatoskr #

![squirrel-image](images/noun-squirrel-1092288_512x512.png)

[What is a ratatoskr?](https://en.wikipedia.org/wiki/Ratatoskr)

> This tool is designed to interact with GitHub and GitLab APIs and send webhooks to your chat application in the event that new releases or commits are made to selected repositories of interest.
> Python script with sqlite3 `tracker.db` file that maintains persistent data.

## Features ##

1. Integration with Rocket.Chat, Discord, Microsoft Teams, and Slack via webhooks.
2. Integration with GitHub API.
3. Integration with GitLab API.
4. Supports HTTP Sessions.
5. Automatic commit, rollback and DB connection closure and error handling.
6. Supports tracking latest releases and latest commits in case repository doesn't use releases.
7. Basic Create, Read, and Update support for sqlite3.
8. Per GitHub API - Uses user-agent header specific to tool.
9. Reproducible builds via `Pipfile.lock` file.

## Supported Webhook Receivers ##

| Platform        | Supported |
|-----------------|-----------|
| Slack           | ✅        |
| Discord         | ✅        |
| Microsoft Teams | ✅        |
| Rocket.Chat     | ✅        |

## Install ##

### Clone repository ###

```shell
git clone https://github.com/axi0m/ratatoskr
cd ratatoskr/
```

### Using pipenv ###

```shell
pipenv install
```

### Using requirements.txt ###

```shell
pip install --user requirements.txt
```

### Using poetry ###

```shell
poetry install
```

## Usage ##

### Environment and Setup ###

  - Minimum Python Version: `3.6`
  - Pipenv Pinned Python Version: `3.9`

1. Ensure `pipenv` is installed, in your path, and you've got a GitHub Personal Access Token and GitLab Personal Access Token exported to environmental variables `GITHUB_TOKEN` and `GITLAB_TOKEN` respectively.
2. Ensure a reference file called `GitHub_Tools_List.csv` is in the current working directory with the URLs of the repositories you want to monitor. Sample is provided here in the repository.
3. Lastly you must export one of the following environment variables depending on your chat application. Alternatively for 1. and 3. you could create your own `.env` file in the directory and populate the required
variables, see the `.env-template` as an example.

   - `ROCKETCHAT_WEBHOOK`
   - `MSTEAMS_WEBHOOK`
   - `SLACK_WEBHOOK`
   - `DISCORD_WEBHOOK`

### Example GitHub_Tools_List.csv ###

```csv
Repo,Category,Currently Used with Approved Aggressor Collection,Added Date
https://github.com/0x09AL/RsaTokenExtractor,.NET Assembly C# Tools,,
https://github.com/0xhexmex/Aggressor-Scripts,aggscripts,,
https://github.com/0xthirteen/CleanRunMRU,.NET Assembly C# Tools,,
https://github.com/0xthirteen/MoveKit,.NET Assembly C# Tools,,
```

### Command-Line Help ###

```shell
pipenv run python3 ratatoskr.py --help
```

![help](images/help_output.png)

### Load repositories ###

1. Create your spreadsheet of GitHub or GitLab URLs to track, if you leave it in Excel format there is a converter script that'll convert to CSV for you, and the wrapper `runner.sh` script is helpful to always run the conversion and then dump the new CSV for use by the script.
2. If an existing `tracker.db` file is not present, it'll automatically create one for you.

```shell
pipenv run python3 ratatoskr.py --load -p discord
```

[![asciicast](https://asciinema.org/a/510231.png)](https://asciinema.org/a/510231)

### First Run ###

You will want to have a Personal Access Token for GitHub and GitLab, and the webhook exported to environment variables before attempting to load or run the script (assuming you are not using the `.env` file).

```shell
export GITHUB_TOKEN='REDACTED'
export GITLAB_TOKEN='REDACTED'
export DISCORD_WEBHOOK='REDACTED'
pipenv run python3 ratatoskr.py --check -p discord
```

[![asciicast](https://asciinema.org/a/510042.png)](https://asciinema.org/a/510042)

![alerts](images/alerts.png)

### Subsequent Runs ###

Use a scheduler of your choice, cron or systemd-timers on Linux/Unix variants or scheduled tasks in Windows.
There is a `runner.sh` shell script that serves as a template for wrapping this script for use in cron for example.

## References ##

- [Python Requests - Sessions](https://docs.python-requests.org/en/master/user/advanced/#session-objects)
- [Python Standard Library - sqlite3](https://docs.python.org/3/library/sqlite3.html)
- [Using Context Managers with sqlite3](https://docs.python.org/2/library/sqlite3.html#using-the-connection-as-a-context-manager)
- [StackOverflow - Commit and Rollback and With statement](https://stackoverflow.com/questions/19522505/using-sqlite3-in-python-with-with-keyword)
- [StackOverflow - Cursor vs Connection object](https://stackoverflow.com/questions/6318126/why-do-you-need-to-create-a-cursor-when-querying-a-sqlite-database)
- [PyNative - CRUD Operations from Python](https://pynative.com/python-sqlite/#h-perform-sqlite-crud-operations-from-python)
- [GitLab API Docs](https://docs.gitlab.com/ee/api/api_resources.html)
- [Python Requests-HTML - Simple Webscraping](https://docs.python-requests.org/projects/requests-html/en/latest/)
- [Python Pipenv](https://pipenv.pypa.io/en/latest/)
- [Reproducible Builds](https://reproducible-builds.org/)
- [Python Rich - Beautiful CLI](https://github.com/willmcgugan/rich)
- [Webhooks](https://en.wikipedia.org/wiki/Webhook)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Exponential Backoff Algorithm
- Rate limit support for GitLab, GitHub
- Add logging
- Add test coverage
- Add check if tracker.db does not exist, automatically create and prep it
- Add verbose parameter for logging level
- Add parameter to --load to overwrite (delete) entries in DB that don't match reference CSV

## [0.5.0] - 2022-07-19

- Test support for Microsoft Teams, Discord and Slack
- Add GitHub actions to perform SAST and verify builds
- Simplify main function

## [0.4.0] - 2022-06-02

### Fixed

- Fix bug in unnecessary messages posted to webhook where commit or release is `False`. [issue #6](https://github.com/axi0m/ratatoskr/issues/6)
- Fix bug in `save_messages()` function so that output file is a valid JSON file. [issue #7](https://github.com/axi0m/ratatoskr/issues/7)
- Fix bug in parsing GitLab project ID using requests HTML package. [issue #3](https://github.com/axi0m/ratatoskr/issues/3)

### Added

- .gitignore to exclude files generated by macOS, VSCode, Python, and tracker.db
- Basic HTTP 429 support for Rocket.Chat webhooks
- Update documentation for README on environment requirements

### Removed

- TODO comments in the code

### Security

- Use environment variables for webhook definition per https://12factor.net/config

## [0.3.1] - 2021-05-26

### Added

- Fix bug in checking section of main() where a newly inserted repo from loading has a False value in commit and release column
- Remove checking as part of loading new repos from reference CSV

## [0.3.0] - 2021-05-07

### Added

- Fix SQL query in `confirm_table()`
- Add rate limit conditional check and sleep for reset period from GitHub API

## [0.2.1] - 2021-05-04

### Added

- Fixed and tested gitlab integration
- `rich` progress bar added to loading and checking routines
- Added newlines to some console printing to account for rich progress bar
- Added conditional check if trying to prep DB and DB already exists with proper schema

## [0.2.0] - 2021-04-29

### Added

- GitLab Integration
- Rich Console Formatting provided by `rich`

## [0.1.0] - 2021-04-23

### Added

- Initial creation


tests/ratatoskr_test.py
import pytest
from .. import ratatoskr

ratatoskr.verify_environment("GITHUB_TOKEN")


src/tests/ratatoskr_test.py
import pytest
from ratatoskr import verify_environment


def test_verify_environment(token):
    token = verify_environment("GITHUB_TOKEN")
    assert isinstance(token, str)


src/ratatoskr/convert_to_csv.py
#!/usr/bin/python
# Reference: https://betterprogramming.pub/using-python-to-convert-worksheets-in-an-excel-file-to-separate-csv-files-7dd406b652d7
# This python script is to extract each sheet in an Excel workbook as a new csv file

import csv
import xlrd
import sys


def ExceltoCSV(excel_file, csv_file):
    """Convert Excel XLS file to CSV"""

    # Create Excel workbook object
    workbook = xlrd.open_workbook(excel_file)

    # Iterate over each tab/sheet name
    for sheet_name in workbook.sheet_names():

        print(f"[-] INFO Processing sheet name - {sheet_name}")

        # Define our worksheet by index
        sh = workbook.sheet_by_index(0)

        with open(csv_file, "w") as csv_fh:
            writetocsv = csv.writer(csv_fh)
            for rownum in range(sh.nrows):
                new_row = []
                for cell in sh.row(rownum):
                    new_row.append(cell.value)
                writetocsv.writerow(new_row)

        print(f"{sheet_name} has been saved at - {csv_file}")


if __name__ == "__main__":
    ExceltoCSV(excel_file=sys.argv[1], csv_file=sys.argv[2])


src/ratatoskr/ratatoskr.py
#!/usr/bin/python3

import argparse
import csv
import json
import os
import requests
import sqlite3 as sl
import sys
import time
import urllib.parse
from datetime import datetime
from dotenv import load_dotenv
from pathlib import Path
from rich.console import Console
from rich.progress import track
from __init__ import __version__
from __init__ import __prog__

# Get the current timestamp
now = datetime.now()
dt_formatted = now.strftime("%d/%m/%Y %H:%M:%S")

# Define header values
USERAGENT = f"ratatoskr-{__version__}"

# Init rich console
console = Console()

# Load .env file
load_dotenv()

# Create filename to save messages if provider is down
# Format YYYY-MM-DD
dt_formatted_filename = now.strftime("%Y-%m-%d")
# Get Process ID
pid = os.getpid()
# Construct filename to save message state
filename = f"ratatoskr_{dt_formatted_filename}_{pid}.json"


def verify_environment(environment_variable):
    """Verify if a provided environment variable has a value and return that value if true"""

    value = os.getenv(environment_variable)
    if not value:
        return None
    else:
        return value


def get_ratelimit_status(session):
    """Get the rate limit status from GitHub API"""

    query_url = "https://api.github.com/rate_limit"
    response = session.get(query_url, timeout=5)

    if response.status_code == 200:
        requests_remaining = int(response.headers["X-RateLimit-Remaining"])
        requests_reset_time = int(response.headers["X-RateLimit-Reset"])
        result = (requests_remaining, requests_reset_time)
        return result
    if response.status_code != 200:
        return None


def get_urls(filename):
    """Read in list of GitHub repositories to monitor releases"""

    # We use a list of tuples instead of a dictionary since the owners (or keys)
    # may be duplicates for different repos by same owner in URL list
    repositories = []

    with open(filename) as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=",")
        line_count = 0
        for row in csv_reader:
            if line_count == 0:
                line_count += 1
                pass
            else:
                components = row[0].split("/")
                owner = components[3]
                repo = components[4]
                if "gitlab" in row[0]:
                    combo = (owner, repo, "gitlab")
                    repositories.append(combo)
                    line_count += 1
                elif "github" in row[0]:
                    combo = (owner, repo, "github")
                    repositories.append(combo)
                    line_count += 1
                else:
                    line_count += 1
        console.print(f"[*] INFO - Processed {line_count} lines.", style="bold green")
        return repositories


def get_gitlab_latest_release(session, repository):
    """Get latest release for given GitLab public project

    session - A requests Session object
    repository - A tuple like ('Owner', 'Repo', 'gitlab')
    """

    # GitLab API will take URL-encoded namespace/project URI component
    project = urllib.parse.quote_plus(f"{repository[0]}/{repository[1]}")
    query_url = f"https://gitlab.com/api/v4/projects/{project}/releases"
    response = session.get(query_url, timeout=5)

    if "json" in response.headers.get("Content-Type"):
        response_json = response.json()

        if response_json == [] and response.status_code == 200:
            console.print(
                f"\n[!] INFO - No release found for project {repository[0]}/{repository[1]}",
                style="bold yellow",
            )
            return None

    if response.status_code == 404:
        console.print(
            f"[!] WARN - Project {repository[0]}/{repository[1]} was not found at {query_url} be sure to confirm the URL",
            style="bold red",
        )

    try:
        latest_release = response_json[0]["_links"].get("self")
    except KeyError:
        return None

    if latest_release:
        return latest_release


def get_gitlab_latest_commit(session, repository):
    """Get latest commit for given GitLab public project

    session - A requests Session object
    repository - A tuple like ('Owner', 'Repo', 'gitlab')
    """

    # GitLab API will take URL-encoded namespace/project URI component
    project = urllib.parse.quote_plus(f"{repository[0]}/{repository[1]}")
    query_url = f"https://gitlab.com/api/v4/projects/{project}/repository/commits"
    response = session.get(query_url, timeout=5)

    if "json" in response.headers.get("Content-Type"):
        response_json = response.json()

        if response.status_code == 404:
            console.print(
                f"[!] WARN - Project {repository[0]}/{repository[1]} was not found at {query_url} be sure to confirm the URL",
                style="bold red",
            )

        latest_commit = response_json[0]
        if latest_commit.get("web_url"):
            return latest_commit["web_url"]
    else:
        return None


def get_latest_release(session, repository):
    """Get the latest GitHub release for given repository

    session - A requests Session object
    repository - A tuple like ('Owner', 'Repo', 'github')
    """

    # Sample input ('outflanknl', 'RedELK', 'github')

    query_url = (
        f"https://api.github.com/repos/{repository[0]}/{repository[1]}/releases/latest"
    )
    response = session.get(query_url, timeout=5)

    if "json" in response.headers.get("Content-Type"):
        response_json = response.json()
        if response_json.get("html_url"):
            return response_json["html_url"]
    else:
        return None


def get_latest_commit(session, repository):
    """Get the latest commit for a given list of repos in ('Owner', 'Repo') format"""

    query_url = f"https://api.github.com/repos/{repository[0]}/{repository[1]}/commits"
    response = session.get(query_url, timeout=5)

    if not response:
        return None

    if "json" in response.headers.get("Content-Type"):
        response_json = response.json()
        latest_commit = response_json[0]
        if latest_commit.get("html_url"):
            return latest_commit["html_url"]
    else:
        return None


def update_tracker(connection, update):
    """Update tracker DB with the latest release and commit

    connection - SQLite DB Connection Object
    update - List of commit, release, datetime, GitHub username, repo name, and whether GitLab or GitHub
    """

    # update_object = [commit, release, dt_formatted, repo[0], repo[1], repo[2]]
    # sample SQL entry = its-a-feature|Mythic|0|https://github.com/its-a-feature/Mythic/commit/75a46ef1c06e58ffaed2b036c3e4adf67b72bbc4|12/05/2021 12:37:32|github

    sql = "UPDATE repo SET latest_commit = ?, latest_release = ?, last_updated = ? WHERE  owner = ? AND repo = ? AND website = ?"

    try:
        cursor = connection.cursor()
        with connection:
            cursor.execute(sql, update)
    except sl.IntegrityError as e:
        console.print(
            f"[!] ERROR - Unable to update repo {update[4]} in tracker DB",
            style="bold red",
        )
        console.print(f"{e}")
        sys.exit(1)


def insert_repo(connection, newrepo):
    """Insert newly identified repository to track

    connection - SQLite DB Connection Object
    newrepo - List of user, repo URL, last updated timestamp, GitHub or GitLab
    """

    # newrepo = [repo[0], repo[1], dt_formatted, repo[2]]

    sql = "insert into repo (owner, repo, last_updated, website) values(?, ?, ?, ?)"

    try:
        cursor = connection.cursor()
        with connection:
            cursor.execute(sql, newrepo)
    except sl.IntegrityError as e:
        console.print(
            f"[!] ERROR - Unable to insert new repo into tracker DB", style="bold red"
        )
        console.print(f"{e}")
        sys.exit(1)


def confirm_table(connection):
    """Verify if repo table has been created

    connection - SQLite DB Connection Object
    """

    cursor = connection.cursor()
    with connection:
        cursor.execute("select * FROM sqlite_master WHERE type='table' and name='repo'")
        data = cursor.fetchall()
        if len(data) == 0:
            return None
        else:
            console.print(f"[+] INFO - Table already exists", style="bold green")
            return True


def delete_repo(connection, repo):
    """Delete repository from tracker db

    connection - SQLite DB Connection Object
    repo - List of owner and repo
    """

    sql = "DELETE FROM repo WHERE owner = ? AND repo = ?"

    try:
        cursor = connection.cursor()
        with connection:
            cursor.execute(sql, repo)
    except sl.IntegrityError as e:
        console.print(
            f"[!] ERROR - Unable to delete repo from tracker DB", style="bold red"
        )
        console.print(f"{e}")
        sys.exit(1)


def confirm_repo(connection, repo):
    """Verify if the owner and repository name is already setup in the tracker database

    connection - SQLite DB Connection Object
    repo - List of owner and repo
    """

    cursor = connection.cursor()
    with connection:
        cursor.execute(
            "select * from repo WHERE owner = ? AND repo = ?",
            [repo[0], repo[1]],
        )
        data = cursor.fetchall()
        if len(data) == 0:
            return None
        else:
            return True


def bootstrap_db(connection):
    """Bootstrap sqlite3 db with REPO table

    connection - SQLite DB Connection Object
    """

    try:
        cursor = connection.cursor()
        with connection:
            cursor.execute(
                "create table repo (owner, repo, latest_release, latest_commit, last_updated, website)"
            )
    except sl.IntegrityError as e:
        console.print(f"[!] ERROR - Unable to create repo table", style="bold red")
        console.print(f"{e}")
        sys.exit(1)


def dump_table(connection):
    """Print the tracker database

    connection - SQLite DB Connection Object
    """

    cursor = connection.cursor()
    with connection:
        data = cursor.execute("select * from repo")
        for row in data:
            print(row)


def read_repositories(connection):
    """Return all repositories in the tracker database

    connection - SQLite DB Connection Object
    """

    repositories = []

    cursor = connection.cursor()
    with connection:
        data = cursor.execute("select * from repo")
        for row in data:
            repositories.append(row)
    return repositories


def save_messages(data, filename):
    """Write messages as JSON to disk in the event webhook is unsuccessful

    data - JSON object
    filename - Filname to write to disk
    """

    try:
        with open(filename, "rt") as fh:
            existing_data = json.load(fh)
    except FileNotFoundError as notfound:
        console.print(
            f"[!] WARN - Filename [blue]{filename}[/blue] does not exist [i]will create[/i]: {notfound}",
            style="bold yellow",
        )
        existing_data = []
    except IOError as e:
        console.print(
            f"[!] ERROR - Unable to read file [blue]{filename}[/blue]: {e}",
            style="bold yellow",
        )
        existing_data = []

    # Update the dict object with new data passed to function
    existing_data.append(data)

    # Write our merged JSON object to disk in the event we have to resend the messages
    with open(filename, "wt") as fh:
        json.dump(existing_data, fh)

    console.print(
        f"[!] WARN - Wrote messages to file [blue]{filename}[/blue]",
        style="bold yellow",
    )


def send_webhook(message, webhook_url, provider, filename):
    """Send web request to webhook URL

    message - Message text for chat notification
    webhook_url - Webhook URL to send request to
    provider - Webhook provider to format message
    filename - Filename to write to disk in event of failure
    """

    # https://docs.microsoft.com/en-us/microsoftteams/platform/webhooks-and-connectors/how-to/add-incoming-webhook
    if provider == "msteams":
        data = {"Text": message}

    # https://api.slack.com/messaging/webhooks
    if provider == "slack":
        data = {"text": message}

    # https://discord.com/developers/docs/resources/webhook
    if provider == "discord":
        data = {"content": message}

    # https://docs.rocket.chat/guides/administration/admin-panel/integrations
    if provider == "rocketchat":
        data = {
            "username": "rocket.cat",
            "icon_emoji": ":chipmunk:",
            "attachments": [{"text": message, "color": "#764FA5"}],
        }

    # HTTP POST to our Webhook URL
    r = requests.post(webhook_url, json=data)

    # Verify successful status code via .ok method on response object
    # Most APIs return 200, Discord returns 204
    if not r.ok:
        console.print(
            f"\n[!] ERROR - Webhook was unsuccessful status code is {r.status_code}: {r.text}",
            style="bold red",
        )
        # Save messages to disk in event we don't succesfully POST
        save_messages(message, filename)

        # Return condition and response object
        return (False, r)

    if r.ok:
        return (True, r)


def parse_arguments():
    """Parse command-line arguments"""

    # Define parser
    parser = argparse.ArgumentParser()

    # Create mutually exclusive group
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument(
        "-l",
        "--load",
        action="store_true",
        help="load the repositories to watch into the database",
    )
    group.add_argument(
        "-c",
        "--check",
        action="store_true",
        help="check for new repository releases and commits",
    )
    group.add_argument(
        "-v", "--version", action="version", version=f"{__prog__} {__version__}"
    )
    group.add_argument(
        "-e",
        "--examples",
        action="store_true",
        help="display usage examples and exit",
    )
    parser.add_argument(
        "-p",
        "--provider",
        type=str,
        choices=["rocketchat", "discord", "msteams", "slack"],
        help="provider to use; required with --check",
    )

    # Parse our arguments into internal variables
    args = parser.parse_args()

    # Cleaner variable names
    load = args.load
    check = args.check
    provider = args.provider

    if provider is None:
        console.print(
            f"[!] ERROR - Chat provider was not provided by [green]--provider[/green] argument!",
            style="bold red",
        )
        sys.exit(1)

    return {"Load": load, "Check": check, "Provider": provider}


def prepare_database(filename):
    """Prepare the database

    filename - SQLite database filename to track repos
    """

    # Check if tracker.db file exists or not
    p = Path(filename)

    if p.exists():
        console.print(f"[+] INFO - [blue]{filename}[/blue] exists", style="bold green")
        # DB Connection
        con = sl.connect(filename, timeout=5)

        # Confirm the table exists, use the connection object
        confirm_result = confirm_table(con)

        # If the table already exists in the tracker.db file
        if confirm_result:
            console.print(
                f"[+] INFO - Tracker database is already prepared", style="bold green"
            )
            return (True, con)

        # If the table does not exist, create it!
        if not confirm_result:
            console.print(
                f"[+] INFO - Preparing database tables in [blue]{filename}[/blue] file..",
                style="bold green",
            )

            # We want to try and handle the sqlite OperationalError condition at least
            try:
                bootstrap_db(con)
                return (True, con)
            except sl.OperationalError as e:
                console.print(
                    f"[!] ERROR - Database has already been initialized!",
                    style="bold red",
                )
                console.print(f"{e}")
                return (False, con)

    # If the file does not exist, create it and prepare it
    if not p.exists():
        console.print(
            f"[!] WARN - [blue]{filename}[/blue] does not exist, creating..",
            style="bold yellow",
        )
        # DB Connection
        con = sl.connect(filename, timeout=5)

        # We want to try and handle the sqlite OperationalError condition at least
        try:
            bootstrap_db(con)
            return (True, con)
        except sl.OperationalError as e:
            console.print(
                f"[!] ERROR - Database has already been initialized!",
                style="bold red",
            )
            console.print(f"{e}")
            return (False, con)


def main():
    """Main function"""

    # Print a pretty header to console
    console.print(f" :chipmunk:  ~ Ratatoskr the Norse Squirrel God ~ :chipmunk:")

    # High-level function to parse arguments
    arguments = parse_arguments()

    # Verify tokens and webhook
    github_token = verify_environment("GITHUB_TOKEN")
    gitlab_token = verify_environment("GITLAB_TOKEN")

    # Exit if we don't have GitHub API token
    if not github_token:
        console.print(
            f"[!] ERROR - No GitHub Personal Access Token in environment variables",
            style="bold red",
        )
        sys.exit(1)

    # Exit if we don't have GitLab API Token
    if not gitlab_token:
        console.print(
            f"[!] ERROR - No GitLab Personal Access Token in environment variables",
            style="bold red",
        )
        sys.exit(1)

    # Parse provider and format for env check
    prefix = arguments["Provider"].upper()

    # Verify that our provider webhook is in the environment
    webhook_url = verify_environment(f"{prefix}_WEBHOOK")

    # Exit if we don't have webhook URL
    if not webhook_url:
        console.print(
            f"[!] ERROR - No webhook URL found in environment variables",
            style="bold red",
        )
        sys.exit(1)

    # Define our headers
    github_custom_headers = {
        "Accept": "application/vnd.github.v3+json",
        "User-Agent": USERAGENT,
        "Authorization": "token {}".format(github_token),
    }
    gitlab_custom_headers = {
        "Accept": "application/vnd.github.v3+json",
        "User-Agent": USERAGENT,
        "PRIVATE-TOKEN": "{}".format(gitlab_token),
    }

    # Create reusable HTTP session object
    s_github = requests.Session()
    s_gitlab = requests.Session()

    # Update Session headers
    s_github.headers.update(github_custom_headers)
    s_gitlab.headers.update(gitlab_custom_headers)

    # Prepare the tracker.db file before loading data
    db_prep_result = prepare_database("tracker.db")

    # Ensure we got True from previous function call
    if not db_prep_result[0]:
        console.print(f"[!] ERROR - Preparing database!", style="bold red")
        sys.exit(1)

    # Use a friendly name for our connection object
    db_connection_handler = db_prep_result[1]

    # Check rate limits
    github_ratelimit_response = get_ratelimit_status(s_github)

    if github_ratelimit_response is None:
        console.print(f"[!] ERROR Unable to confirm rate limits", style="bold red")
        sys.exit(1)

    # If user provided --load argument, read CSV and load into tracker
    if arguments["Load"]:
        # Extract all the URLs from the first column in the CSV
        repositories = get_urls("GitHub_Tools_List.csv")

        console.print(
            f"[+] Loading repositories to monitor into tracker..", style="bold green"
        )

        # We enumerate over all repository URLs
        for repo in track(
            sequence=repositories,
            description="Loading...",
            update_period=1.0,
            auto_refresh=False,
        ):
            # Check if already tracking in database
            confirmation = confirm_repo(db_connection_handler, repo)

            # Verify that repository has been inserted into tracker
            if confirmation:
                pass
            # If it has not been inserted, load it in
            elif not confirmation:
                # Prepare our SQL insert
                newrepo = [repo[0], repo[1], dt_formatted, repo[2]]
                # Perform insert by passing our DB handler and new repository
                insert_repo(db_connection_handler, newrepo)
        sys.exit(0)

    # If user provided --check argument, read all repos from DB
    if arguments["Check"]:
        # Read tracker.db and populate all our repositories in memory
        repositories = read_repositories(db_connection_handler)

        for count, repo in enumerate(
            track(
                sequence=repositories,
                description="Checking repositories...",
                update_period=1.0,
            )
        ):
            if repo[5] == "github":
                # Get latest release URL
                release = get_latest_release(s_github, repo)
                # Get latest commit URL
                commit = get_latest_commit(s_github, repo)
            if repo[5] == "gitlab":
                # Get latest release URL
                release = get_gitlab_latest_release(s_gitlab, repo)
                # Get latest commit URL
                commit = get_gitlab_latest_commit(s_gitlab, repo)

            # Check if latest release matches DB
            if repo[2] != release and release is not None:
                console.print(
                    f"\n[+] New release for repository {repo[1]}: {release}",
                    style="bold green",
                )

                # Update the database
                update = [commit, release, dt_formatted, repo[0], repo[1], repo[5]]
                update_tracker(db_connection_handler, update)

                # Send notification to webhook
                message = f"New release for repository {repo[1]}: {release}"
                response = send_webhook(
                    message, webhook_url, arguments["Provider"], filename
                )

                # If response code is 429, backoff
                if response[1].status_code == 429:
                    delay_time = 60
                    console.print(
                        f"[!] WARN - Too many requests, backing off for [blue]{delay_time}[/blue] seconds"
                    )
                    time.sleep(delay_time)
                    response = send_webhook(
                        message, webhook_url, arguments["Provider"], filename
                    )

            if repo[3] != commit and commit is not None:
                console.print(
                    f"\n[+] New commit for repository {repo[1]}: {commit}",
                    style="bold green",
                )

                # Update the database
                update = [commit, release, dt_formatted, repo[0], repo[1], repo[5]]
                update_tracker(db_connection_handler, update)

                # Send notification to webhook
                message = f"New commit for repository {repo[1]}: {commit}"
                response = send_webhook(
                    message, webhook_url, arguments["Provider"], filename
                )

                # If response code is 429, backoff
                if response[1].status_code == 429:
                    delay_time = 60
                    console.print(
                        f"[!] WARN - Too many requests, backing off for [blue]{delay_time}[/blue] seconds"
                    )
                    time.sleep(delay_time)
                    response = send_webhook(
                        message, webhook_url, arguments["Provider"], filename
                    )


if __name__ == "__main__":
    main()
